feat: string and message value renderers for textual

JimLarson opened the pull request:
## Description

Closes: #12713 
Refs: #12878

Sign mode textual value renderers for string and message.

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

JimLarson commented:
Draft checkin to illustrate issue in `valuerenderer.go:GetValueRenderer()` for message types. Can't get to the Go structure from descriptors.

peterbourgon commented on docs/architecture/adr-050-sign-mode-textual.md:
```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
+or needs to be present only for signature integrity (see below).
 
-The encoding and decoding operations between a Protobuf transaction (whose definition can be found [here](https://github.com/cosmos/cosmos-sdk/blob/master/proto/cosmos/tx/v1beta1/tx.proto#L13)) and the string array MUST be bijective.
+### Invertible Rendering
```
What motivates the change in nomenclature away from "bijective"?

JimLarson replied:
Note that this draft PR is based off of the not-yet-committed #13434, so you're commenting on changes that will go away once this PR is rebased. You comments might be more appropriate for #13434.

But to quickly answer, "bijective" is not exactly correct. That would require that every valid Tx round-trip back to an equal Tx (yes) and also that every possible string round-trip back to an equal string (no). Having just the first condition makes `Format()` an "injection" (i.e. one-to-one), but this was deemed to be too technical, hence the use of "invertible".

peterbourgon replied:
Should I move my comments to #13434, or is there enough visibility here?

amaurym replied:
yes that's preferable. Let's keep this PR as message value renderer only

JimLarson replied:
Hmm - #13434 (structured screens) is just awaiting final approval, so I don't want to derail that. I'm not seeing anything in these comments that would make me want to add any late changes. This discussion seems more appropriate for #12910 - but as you can see that ship sailed a while ago. @peterbourgon it would be most appropriate if you could open an issue for discussion and raise you concerns there. I'm going to change the target for this draft to we can see the actual content of this PR better.

peterbourgon commented on docs/architecture/adr-050-sign-mode-textual.md:
````diff
@@ -181,25 +229,9 @@ When this option is set on a `Msg`, a registered function will transform the `Ms
-}
-pass_check()
-```
+Moreover, the renderer must provide 2 functions: one for formatting from Protobuf to string, and one for parsing string to Protobuf. These 2 functions are provided by the application developer. To satisfy point #1, these 2 functions MUST be bijective with each other. Bijectivity of these 2 functions will not be checked by the SDK at runtime. However, we strongly recommend the application develope ...
````
If the application fails to ensure bijectivity, are any guarantees asserted by Cosmos and/or Tendermint violated?

peterbourgon commented on docs/architecture/adr-050-sign-mode-textual.md:
```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
 
-- by providing a set of test fixtures between a transaction's Proto JSON representation and its TEXTUAL representation, and checking that encoding/decoding in both directions matches the fixtures,
-- by using property testing on the proto transaction itself, and testing that the composition of encoding and decoding yields the original transaction itself.
+Note that the existence of an inverse function ensures that the
+rendered text contains the full information of the original transaction,
+not a hash or subset.
```
Well, it claims this property, but doesn't really ensure it, right?

peterbourgon commented on docs/architecture/adr-050-sign-mode-textual.md:
```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
-- The initial `*` character is optional and denotes the Ledger Expert mode, see #5.
-- Strings can also include a number of `>` character to denote nesting.
-- In the case where the first Regex capture group is not empty, it represents an indicative key, whose associated value is given in the second capture group. This MAY be used in the Ledger app to perform custom on-screen formatting, for example to break long lines into multiple screens.
+The correctness and security of `SIGN_MODE_TEXTUAL` is guaranteed by demonstrating an inverse function from the rendering to transaction protos.
+This means that it is impossible for a different protocol buffer message to render to the same text.
```
It appears that inversability is not guaranteed or even verified by the SDK, right? It's just assumed to be true. So if an implementation happens to have a bug that violates this assumption, and isn't fully inversible in some circumstance, what happens?

peterbourgon commented on docs/architecture/adr-050-sign-mode-textual.md:
```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
+### Transaction Hash Malleability
 
-The `<value>` itself can contain the `": "` characters.
+When client software forms a transaction, the "raw" transaction (`TxRaw`) is serialized as a proto
+and a hash of the resulting byte sequence is computed.
```
Note that [protobuf serialization is not deterministic](https://developers.google.com/protocol-buffers/docs/encoding).

> By default, repeated invocations of serialization methods on the same protocol buffer message instance may not produce the same byte output. That is, the default serialization is not deterministic.
> 
> Deterministic serialization only guarantees the same byte output for a particular binary. The byte output may change across different versions of the binary.

amaurym commented on tx/textual/valuerenderer/message.go:
```diff
@@ -0,0 +1,154 @@
+	fds     []protoreflect.FieldDescriptor
+}
+
+func NewMessageValueRenderer(t *Textual, msgType protoreflect.ProtoMessage) ValueRenderer {
```
I would probably have done:

```suggestion
func NewMessageValueRenderer(t *Textual, msgDesc protoreflect.MessageDescriptor) ValueRenderer {
```

amaurym replied:
And then the `protoreflect.FullName` -> `protoreflect.Message` resolving happens inside `Parse`, instead of inside valuerenderer.go

JimLarson replied:
Yeah, that's what I was thinking after dabbling with Any. Tried it, and it works quite well.

JimLarson commented:
@joeabbey Thanks for finding al those problems in the docs! Unfortunately, those aren't really part of this PR - the structured screens PR hadn't been merged yet, so I branched this off the end of it, and its contents were displayed here since there was no way to make it the target. In any case, I've created #13608 for minor documentation fixes and addressed your comments there.

JimLarson commented on docs/architecture/adr-050-sign-mode-textual-annex2.md:
```diff
@@ -0,0 +1,122 @@
+## Context
+
+`SIGN_MODE_TEXTUAL` allows a legible version of a transaction to be signed
+on a hardware security devices, such as a Ledger. Early versions of the
```
Done (see #13608).

JimLarson commented on docs/architecture/adr-050-sign-mode-textual-annex2.md:
```diff
@@ -0,0 +1,122 @@
+- Expert mode screens are only presented if the device is in expert mode.
+
+- Each line of the screen starts with a number of `>` characters equal
+to the screen's indetation level, followed by a `+` character if this
```
Done.

JimLarson commented on docs/architecture/adr-050-sign-mode-textual-annex2.md:
```diff
@@ -0,0 +1,122 @@
+or if this header is followed by a `>`, `+`, or space.
+
+- If the line ends with whitespace or an `@` character, and `@` character
+is appended to the line.
```
Done.

JimLarson commented on docs/architecture/adr-050-sign-mode-textual.md:
```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
-Bijectivity will be tested in two ways:
+Note that this inverse function does not need to perform correct
+parsing or error signaling for the whole domain of textual data.
+Merely that the range of valid transactions be invertible under
```
Still considering this. Where'd I put my Strunk & White?

JimLarson commented on docs/architecture/adr-050-sign-mode-textual.md:
```diff
@@ -23,72 +24,119 @@ Protobuf-based SIGN_MODE_DIRECT was introduced in [ADR-020](./adr-020-protobuf-t
 
-- by providing a set of test fixtures between a transaction's Proto JSON representation and its TEXTUAL representation, and checking that encoding/decoding in both directions matches the fixtures,
-- by using property testing on the proto transaction itself, and testing that the composition of encoding and decoding yields the original transaction itself.
+Note that the existence of an inverse function ensures that the
```
I prefer the original, but push back if you feel strongly.

amaurym commented on tx/textual/internal/testdata/message.json:
```diff
@@ -0,0 +1,139 @@
+            },
+            "bar": {}
+        },
+        "screens": [
+            {"text": "Foo object"},
+            {"text": "Full name: subfield", "indent": 1},
+            {"text": "Left: Foo object", "indent": 1},
+            {"text": "Right: Foo object", "indent": 1},
+            {"text": "Nickname: junior", "indent": 2},
+            {"text": "Bar: Bar object", "indent": 1}
+        ]
```
This is an interesting test case. I'm wondering if the indentation is enough to clearly denote for humans the beginning and end of a struct. From a bijectivity view, yes `indent` is enough, but from a human-readability one, it's almost as if we're expecting the next line to be Bar's fields.

When writing the spec initially, I was pondering on whether to add a `End of Foo object` (like we do in `repeated`) or not. Looking at this test case, I'm considering yes.

JimLarson replied:
I'd weigh the utility of the footer screen against the disadvantage of additional clutter. I think it's important to optimize legibility for the common cases, and I doubt that present-but-empty messages will be common. This was mostly meant to explore the corner cases of the renderer than to represent an important test case.

amaurym commented on tx/textual/internal/testdata/message.json:
```diff
@@ -0,0 +1,139 @@
+            {"text": "Bar id: quux", "indent": 3},
+            {"text": "Data: FFFE", "indent": 3}
+        ]
+    }
```
I would add a couple of parsing error cases:
- `[]`
- `[{"text": "Bar object"}]`
- `[{"text": "Foo object", "indent": 1}]`
- `[{"text": "Foo object"}, {"text": "Full name: deep", indent: 2}]`
- `[{"text": "Foo object"}, {"text": "Leftover screen"}]`

JimLarson replied:
I'm inclined to not care about the limits of `Parse()`. It must successfully invert any Screens produced by `Format()`, but other than that it can really do whatever it wants.

amaurym reviewed (commented):
Nice work on this PR @JimLarson 👏 

I left 2 small comments. I'll approve and put automerge once they are addressed

joeabbey commented on tx/textual/valuerenderer/valuerenderer.go:
```diff
@@ -75,6 +75,9 @@ func (r Textual) GetValueRenderer(fd protoreflect.FieldDescriptor) (ValueRendere
 			return NewIntValueRenderer(), nil
 		}
 
+	case fd.Kind() == protoreflect.StringKind:
+		return stringValueRenderer{}, nil
```
Do we need two StringKind cases?

why not incorporate in the above?
```
case fd.Kind() == protoreflect.StringKind:
if proto.GetExtension(fd.Options(), cosmos_proto.E_Scalar) == "" {
		return stringValueRenderer{}, nil
} else {
		scalar, ok := proto.GetExtension(fd.Options(), cosmos_proto.E_Scalar).(string)
...
```

JimLarson commented:
Updated. @amaurym please re-enable automerge.
